tensorflow_probability/python/mcmc/replica_exchange_mc.py

```diff
@@ -0,0 +1,365 @@
+
+  tfd = tf.contrib.distributions
+
+  # Tuning acceptance rates:
+  dtype = np.float32
+  num_warmup_iter = 1000
+  num_chain_iter = 1000
+
+  x = tf.get_variable(name='x', initializer=np.zeros(2, dtype=dtype))
+
+  # Target distribution is Mixture Normal.
+  target = tfd.MixtureSameFamily(
+      mixture_distribution=tfd.Categorical(probs=[0.5, 0.5]),
+      components_distribution=tfd.MultivariateNormalDiag(
+          loc=[[-5., -5], [5., 5.]],
+          scale_identity_multiplier=[1., 1.]))
+
+  # Initialize the ReplicaExchangeMC sampler.
+  remc = tfp.mcmc.ReplicaExchangeMC(
+      target_log_prob_fn=target.log_prob,
+      inverse_temperatures=tf.pow(10., tf.linspace(0., -2., 5)),
+      replica_kernel_class=tfp.mcmc.HamiltonianMonteCarlo,
+      step_size=0.5,
+      num_leapfrog_steps=3)
+
+  # One iteration of the ReplicaExchangeMC
+  init_results = remc.bootstrap_results(x)
+  next_x, other_results = remc.one_step(
+      current_state=x,
+      previous_kernel_results=init_results)
+
+  x_update = x.assign(next_x)
+  replica_update = [init_results.replica_states[i].assign(
+      other_results.replica_states[i]) for i in range(remc.n_replica)]
+
+  warmup = tf.group([x_update, replica_update])
+
+  init = tf.global_variables_initializer()
+
+  with tf.Session() as sess:
+    sess.run(init)
+    # Warm up the sampler
+    for _ in range(num_warmup_iter):
+      sess.run(warmup)
+    # Collect samples
+    samples = np.zeros([num_chain_iter, 2])
+    replica_samples = np.zeros([num_chain_iter, 5, 2])
+    for i in range(num_chain_iter):
+      _, x_, replica_x_ = sess.run([x_update, x, replica_update])
+      samples[i] = x_
+      replica_samples[i] = replica_x_
+
+  plt.figure(figsize=(8, 8))
+  plt.xlim(-10, 10)
+  plt.ylim(-10, 10)
+  plt.plot(samples[:, 0], samples[:, 1], '.')
+  plt.show()
+  ```
+
+  """
+  def __init__(self, target_log_prob_fn, inverse_temperatures,
+               replica_kernel_class,
+               exchange_proposed_fn=default_exchange_proposed_fn(1.),
+               seed=None, name=None, **kwargs):
+    """Instantiates this object.
+
+    Args:
+      target_log_prob_fn: Python callable which takes an argument like
+        `current_state` (or `*current_state` if it's a list) and returns its
+        (possibly unnormalized) log-density under the target distribution.
+      inverse_temperatures: sequence of inverse temperatures to perform
+        samplings with each replica.
+      replica_kernel_class: mcmc kernel class for multiple sampling with
+        different temperatures in parallel
+      exchange_proposed_fn: Python callable which take a number of replicas, and
+        return combinations of replicas for exchange.
+      seed: Python integer to seed the random number generator.
+      name: Python `str` name prefixed to Ops created by this function.
+        Default value: `None` (i.e., "remc_kernel").
+      **kwargs: Arguments for `replica_kernel_class`.
+    """
```

Comment:
raise ValueError if `inverse_temperatures.shape.ndims is None or inverse_temperatures.shape[0].value is None`